src/classes/CW_CampaignTriggerHandlerTest.cls

```diff
@@ -0,0 +1,266 @@
+		
+		// assert deleted campaign (child) points are not rolled up to roadmap (parent)
+		roadmaps = [select Id, Points_Used__c from CW_Roadmap__c where Id = :roadmap.Id];
+		system.assertEquals(40.0, roadmaps[0].Points_Used__c);
+		
+		// undelete Campaign
+		undelete camps[0];
+		
+		// assert undeleted campaign (child) points are included in roll up to roadmap (parent)
+		roadmaps = [select Id, Points_Used__c from CW_Roadmap__c where Id = :roadmap.Id];
+		system.assertEquals(60.0, roadmaps[0].Points_Used__c);
+		
+	}
+	
+	// test that points on Campaign (child) are rolled up to Roadmap (parent) when lookup to Roadmap is removed from a Campaign record
+	@isTest static void testCampaignRollupSumOnRoadmapWhenLookupToRoadmapRemoved() {
+		
+		// setup data
+		CW_Roadmap__c roadmap = TestDataUtil.getRoadmap();
+		insert roadmap;
+		
+		List<CW_Campaign__c> camps = TestDataUtil.getCampaigns(2);
+		for (CW_Campaign__c c :camps) {
+			c.Roadmap__c = roadmap.Id;
+			c.Points__c = 5.0;
+		}
+		
+		// insert Campaign (with points)
+		insert camps;
+		
+		// assert inserted campaigns (child) points are rolled up to roadmap (parent)
+		List<CW_Roadmap__c> roadmaps = [select Id, Points_Used__c from CW_Roadmap__c where Id = :roadmap.Id];
+		system.assertEquals(10.0, roadmaps[0].Points_Used__c);
+		
+		// update Campaign to remove lookup to Roadmap
+		camps[0].Roadmap__c = null;
+		update camps[0];
+		
+		// assert campaign (child which does not have a Roadmap) points are not rolled up to roadmap (parent)
+		roadmaps = [select Id, Points_Used__c from CW_Roadmap__c where Id = :roadmap.Id];
+		system.assertEquals(5.0, roadmaps[0].Points_Used__c);
+		
+		// update Campaign to add lookup to Roadmap
+		camps[0].Roadmap__c = roadmap.Id;
+		update camps[0];
+		
+		// assert campaign (child) points are rolled up to roadmap (parent)
+		roadmaps = [select Id, Points_Used__c from CW_Roadmap__c where Id = :roadmap.Id];
+		system.assertEquals(10.0, roadmaps[0].Points_Used__c);
+	}
+	
+	@isTest static void updateChildSchedulesWithCampaignStageTest() {
+		
+		//setup data
+		List<CW_Campaign__c> camps = TestDataUtil.getCampaigns(2);
+		for (CW_Campaign__c c :camps) c.Stage__c = 'stage 1';
+		
+		// insert Campaign (with stage)
+		insert camps;
+		
+		List<CW_Schedule__c> schedules = new List<CW_Schedule__c>();
+		List<CW_Schedule__c> schedules1 = TestDataUtil.getSchedules(camps[0].Id, 5);
+		List<CW_Schedule__c> schedules2 = TestDataUtil.getSchedules(camps[1].Id, 5);
+		schedules.addAll(schedules1);
+		schedules.addAll(schedules2);
+		
+		// insert schedules
+		insert schedules;
+		
+		// get schedule ids
+		Set<Id> scheduleIds = new Set<Id>();
+		for (CW_Schedule__c s :schedules) scheduleIds.add(s.Id);
+		
+		// update schedules
+		for (CW_Campaign__c c :camps) c.Stage__c = 'stage 2';
+		update camps;
+		
+		// assert schedules have been updated
+		List<CW_Schedule__c> updatedSchedules = [select Id, Campaign_Stage__c from CW_Schedule__c where Id in:scheduleIds];
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.